src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
-f.to_continuous_map.homotopic_rel g.to_continuous_map (cube.boundary n)
+instance inhabited : inhabited (gen_loop N x) := ⟨const⟩
+
+/-- The "homotopy relative to boundary" relation between `gen_loop`s. -/
+def homotopic (f g : gen_loop N x) : Prop := f.1.homotopic_rel g.1 (cube.boundary N)
 
 namespace homotopic
 section
-variables {f g h : gen_loop n x}
+variables {f g h : gen_loop N x}
 
-@[refl] lemma refl (f : gen_loop n x) : homotopic f f := continuous_map.homotopic_rel.refl _
+@[refl] lemma refl (f : gen_loop N x) : homotopic f f := continuous_map.homotopic_rel.refl _
 
-@[symm] lemma symm (H : f.homotopic g) : g.homotopic f := H.symm
+@[symm] lemma symm (H : homotopic f g) : homotopic g f := H.symm
 
-@[trans] lemma trans (H0 : f.homotopic g) (H1 : g.homotopic h) : f.homotopic h := H0.trans H1
+@[trans] lemma trans (H0 : homotopic f g) (H1 : homotopic g h) : homotopic f h := H0.trans H1
 
-lemma equiv : equivalence (@homotopic X _ n x) :=
+lemma equiv : equivalence (@homotopic X _ N x) :=
 ⟨homotopic.refl, λ _ _, homotopic.symm, λ _ _ _, homotopic.trans⟩
 
-instance setoid (n : ℕ) (x : X) : setoid (gen_loop n x) := ⟨homotopic, equiv⟩
+instance setoid (N) (x : X) : setoid (gen_loop N x) := ⟨homotopic, equiv⟩
 
 end
 end homotopic
 
+section
+
+variable [decidable_eq N]
+
+/-- Path from a generalized loop by currying `cube N` into `I × {j // j ≠ i} → I`. -/
+@[simps] def to_path (i : N) : gen_loop N x → Ω (gen_loop {j // j ≠ i} x) const := λ p,
+{ to_fun := λ t, ⟨(p.val.comp (cube.insert_at i).to_continuous_map).curry t,
+    λ y yH, p.property (cube.insert_at i (t, y)) (cube.insert_at_boundary i $ or.inr yH)⟩,
+  continuous_to_fun := by continuity,
+  source' := by { ext t, refine p.property (cube.insert_at i (0, t)) ⟨i, or.inl _⟩, simp },
+  target' := by { ext t, refine p.property (cube.insert_at i (1, t)) ⟨i, or.inr _⟩, simp } }
+
+/-- Generalized loop from a path by uncurrying `I × {j // j ≠ i} → I` into `cube N`. -/
+@[simps] def from_path (i : N) : Ω (gen_loop {j // j ≠ i} x) const → gen_loop N x :=
+λ p, ⟨(⟨λ t, (p t).1, by continuity⟩ : C(I, C(cube _, X))).uncurry.comp
+  (cube.split_at i).to_continuous_map,
+begin
+  rintros y ⟨j, Hj⟩,
+  simp only [subtype.val_eq_coe, continuous_map.comp_apply, to_continuous_map_apply,
+    fun_split_at_apply, continuous_map.uncurry_apply, continuous_map.coe_mk,
+    function.uncurry_apply_pair],
+  by_cases Heq : j = i,
+  { subst Heq, cases Hj; rw Hj; simp only [p.source, p.target]; convert const_eq },
+  { exact gen_loop.boundary _ _ ⟨⟨j, Heq⟩, Hj⟩ },
+end⟩
+
+lemma to_from (i : N) (p : Ω (gen_loop {j // j ≠ i} x) const) : to_path i (from_path i p) = p :=
+begin
+  simp_rw [to_path, from_path, continuous_map.comp_assoc, to_continuous_map_as_coe,
+    to_continuous_map_comp_symm, continuous_map.comp_id], ext, refl,
+end
+
+/-- The `n+1`-dimensional loops are equivalent to the loop space at `const` as a `n`-dimensional
+  loop, defined abtractically for any nonempty index type `N`. -/
```

Comment:
```suggestion
/-- The `n+1`-dimensional loops are in bijection with the loops in the space of
  `n`-dimensional loops with base point `const`.
  We allow an arbitrary indexing type `N` in place of `fin n` here. -/
```
(We may also put the last sentence in the module docstring so we don't need to repeat it in every declaration. BTW, the adverb of "abstract" is "abstractly".)